src/iop/hlrecovery.c

```diff
@@ -538,46 +778,154 @@ static void process_recovery(dt_dev_pixelpipe_iop_t *piece, const void *const iv
+  }
+
+  if(do_recovery)
+  {
+    const float max_distance = dt_image_distance_transform(NULL, distance, pwidth, pheight, 1.0f, DT_DISTANCE_TRANSFORM_NONE);
+
+    if(max_distance > 4.0f)
+    {
+      segmentize_plane(&isegments[DT_IO_PLANNE_ALL], pwidth, pheight);
+      initial_gradients(pwidth, pheight, luminance, distance, recout);
+      dt_masks_extend_border(recout, pwidth, pheight, HLBORDER);
+
+      // now we check for significant all-clipped-segments and reconstruct data
+      for(int id = 2; id < isegments[DT_IO_PLANNE_ALL].nr+2; id++)
+      {
+        const float seg_dist = segment_maxdistance(pwidth, pheight, distance, &isegments[DT_IO_PLANNE_ALL], id);
+        isegments[DT_IO_PLANNE_ALL].val1[id] = seg_dist;
+        
+        if(isegments[DT_IO_PLANNE_ALL].val1[id] > 3.0f)
+          segment_gradients(pwidth, pheight, distance, recout, tmp, recovery_mode, &isegments[DT_IO_PLANNE_ALL], id, seg_border);
+      }
+      dt_masks_blur_fast(recout, gradient, pwidth, pheight, 1.5f, 1.0f, 2.0f);
+      // possibly add some noise
+      const float noise_level = data->noise_level / fmaxf(piece->iscale / roi_in->scale, 1.0f);
+      if(noise_level > 0.0f)
+      {
+        for(int id = 2; id < isegments[DT_IO_PLANNE_ALL].nr+2; id++)
+        {
+          if(isegments[DT_IO_PLANNE_ALL].val1[id] > 3.0f)
+            add_poisson_noise(pwidth, pheight, gradient, &isegments[DT_IO_PLANNE_ALL], id, noise_level);
+        }
+      }
+
+      const float dshift = 2.0f + (float)recovery_closing[recovery_mode];
+#ifdef _OPENMP
+  #pragma omp parallel for default(none) \
+  dt_omp_firstprivate(gradient, distance) \
+  dt_omp_sharedconst(pheight, pwidth, strength, dshift) \
+  schedule(static) collapse(2)
+#endif
+      for(int row = HLBORDER; row < pheight - HLBORDER; row++)
+      {
+        for(int col = HLBORDER; col < pwidth - HLBORDER; col++)
+        {
+          const size_t i = row * pwidth + col;
+          gradient[i] *= strength / (1.0f + expf(-(distance[i] - dshift)));
+        }
+      }
+
+#ifdef _OPENMP
+  #pragma omp parallel for default(none) \
+  dt_omp_firstprivate(out, gradient, max_correction) \
+  dt_omp_sharedconst(width, height, pwidth, p_off) \
+  schedule(static) collapse(2)
+#endif
+      for(int row = 0; row < height; row++)
+      {
+        for(int col = 0; col < width; col++)
+        {
+          const size_t i = (row/2)*pwidth + (col/2) + p_off;
+          const size_t o = row * width + col;
+          out[o] += max_correction * gradient[i];
+        }
+      }
+    }
+  }
+
   dt_get_times(&time3);
 
   if(vmode && fullpipe)
   {
 #ifdef _OPENMP
   #pragma omp parallel for default(none) \
-  dt_omp_firstprivate(out, in, plane, isegments, cmask) \
+  dt_omp_firstprivate(out, in, isegments, cmask, gradient) \
   dt_omp_sharedconst(width, height, pwidth, p_off, filters, vmode) \
-  schedule(static)
+  schedule(static) collapse(2)
 #endif
-    for(size_t row = 0; row < height; row++)
```

Comment:
don't we want to keep all the `size_t` here? Some time ago we have ensured that pixel ref was always a size_t to ensure that very large file could be processed.

Reply:
yes again